lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -5220,28 +5226,93 @@ ParseResult FIRCircuitParser::parseLayerName(SymbolRefAttr &result) {
   return success();
 }
 
-ParseResult FIRCircuitParser::parseOptionalEnabledLayers(ArrayAttr &result) {
-  if (getToken().getKind() != FIRToken::kw_enablelayer) {
-    result = ArrayAttr::get(getContext(), {});
-    return success();
+ParseResult
+FIRCircuitParser::parseOptionalModuleLayerSpec(ArrayAttr &enabledLayers) {
+  enabledLayers = nullptr;
+
+  SmallVector<Attribute> enabledLayersBuffer;
+  while (true) {
+    auto tokenKind = getToken().getKind();
+    // Parse an enablelayer spec.
+    if (tokenKind == FIRToken::kw_enablelayer) {
+      if (parseEnableLayerSpec(enabledLayersBuffer))
+        return failure();
+      continue;
+    }
+    // Didn't parse a layer spec.
+    break;
   }
 
-  if (requireFeature({4, 0, 0}, "modules with layers enabled"))
-    return failure();
+  if (enabledLayersBuffer.size() != 0)
+    if (requireFeature({4, 0, 0}, "modules with layers enabled"))
+      return failure();
+
+  enabledLayers = ArrayAttr::get(getContext(), enabledLayersBuffer);
+  return success();
+}
+
+ParseResult
+FIRCircuitParser::parseOptionalExtModuleLayerSpec(ArrayAttr &enabledLayers,
+                                                  ArrayAttr &knownLayers) {
+  enabledLayers = nullptr;
+  knownLayers = nullptr;
+
+  SmallVector<Attribute> enabledLayersBuffer;
+  SmallVector<Attribute> knownLayersBuffer;
+  while (true) {
+    auto tokenKind = getToken().getKind();
+    // Parse an enablelayer spec.
+    if (tokenKind == FIRToken::kw_enablelayer) {
+      if (parseEnableLayerSpec(enabledLayersBuffer))
+        return failure();
+      continue;
+    }
+    // Parse a knownlayer spec.
+    if (tokenKind == FIRToken::kw_knownlayer) {
+      if (parseKnownLayerSpec(knownLayersBuffer))
+        return failure();
+      continue;
+    }
+    // Didn't parse a layer spec.
+    break;
+  }
 
-  SmallVector<Attribute> layers;
+  if (enabledLayersBuffer.size() != 0)
+    if (requireFeature({4, 0, 0}, "extmodules with layers enabled"))
+      return failure();
+
+  if (knownLayersBuffer.size() != 0)
+    if (requireFeature({5, 0, 0}, "extmodules with known layers"))
```

Comment:
5.0.0 is already released. The next unreleased minor version is: 5.2.0.

Reply:
Actually... should I be using 5.1.0, which is the nextFIRVersion, and also the exportFIRVersion? In order to test the FIRExporter changes I have made, the exportFIRVersion should allow for knownlayer decls. Or do we need to bump the nextFIRVersion?

https://github.com/llvm/circt/blob/main/include/circt/Dialect/FIRRTL/FIRParser.h#L127

Reply:
Using nextFIRVersion!